### ISSUE TITLE
windows add winrepo support for "*.msu" files needing "wusa.exe" (works much like "msiexec.exe")

### ISSUE THREAD

**Issue by TheBigBear** [saltstack/salt#27193]
This is a question, so I can hopefully help add a new feature to windows salt winrepo. 

@twangboy @jfindlay or @terminalmage can one of you point me in the right direction? 

And @UtahDave what do you think about adding ```wusa``` installer support and changing ```msiexec``` for something more generic as mentioned further down? 

The current windows winrepo supports ```msiexec``` to install ```*.msi``` files, and native ```*.exe``` executables as ```installers```. (are there any other installers supported that I am not aware off?) 

In newer versions of windows (not that "new" it's actually been there ever since Vista) they also have another packaging version that ships files in "*.msu" files. The newer versions (or upgrades of) Powershell are distributd that way for example. 

I am trying to find where and how the current ```msiexec.exe``` support is in the ```salt``` windows side and hoping to be able to add ```wusa.exe``` support in.

But I can't for the live of me figure out where the ```msiexec``` functionality is implemented? 

For example to install a newer version of Powershell on windows 7 all that is needed is running ```wusa.exe /quiet Windows6.1-KB2819745-x64-MultiPkg.msu /norestart```

@eliasp had also already asked for this in the lead-up to Berryllium in his comment https://github.com/saltstack/salt/issues/20737#issuecomment-78672235, I believe if one of you can point me in the right direction with regards to where current ```msiexec``` support is and where it's definitions live I can hopefully add the same for ```wusa```?

I also like his more flexible and expandable suggestion of swapping the line ```msiexec = True``` for a more generic ```method = < native | msiexec | powershell | pip | wusa | etc etc >```

**Comment by jfindlay** [comment]
@TheBigBear, thanks for the report.  I think @twangboy is going to be the best person to advise here.

**Comment by alxbse** [comment]
@TheBigBear The msiexec functionality is implemented in the execution module win_pkg and the function install around line 715. Seems like an easy task to add support for wusa as well.

```
# Build the install command
cmd = []
if msiexec:
    cmd.extend(['msiexec', '/i'])
cmd.append(cached_pkg)
cmd.extend(shlex.split(install_flags))
if msiexec and all_users:
    cmd.append('ALLUSERS="1"')
```

**Comment by TheBigBear** [comment]
@alxbse Thanks. I have had a go at it here https://github.com/saltstack/salt/pull/27224, but i am missing something? Can you see what I am missing?

**Comment by alxbse** [comment]
I haven't kept up with the jinja winrepo support changes but my immediate guess is that none of the ifs are true. Maybe add a generic else as a catch-all to the jinja statements?

For example
```
{% if salt['pkg.compare_versions'](grains['osversion'], '==', '6.2') %}
installer: 'http://download.microsoft.com/download/3/F/D/3FD04B49-26F9-4D9A-8C34-4533B9D5B020/W2K12-KB3066438-x64.msu'
{% elif salt['pkg.compare_versions'](grains['osversion'], '==', '6.1') %}
installer: 'http://download.microsoft.com/download/3/F/D/3FD04B49-26F9-4D9A-8C34-4533B9D5B020/Win7AndW2K8R2-KB3066439-x64.msu'
{% elif salt['pkg.compare_versions'](grains['osversion'], '==', '6.3') %}
installer: 'http://download.microsoft.com/download/3/F/D/3FD04B49-26F9-4D9A-8C34-4533B9D5B020/Win8.1AndW2K12R2-KB3066437-x64.msu'
{% else %}
installer: 'my-fallback-that-never-should-happen.exe'
{% endif %}

**Comment by TheBigBear** [comment]
@alxbse you are a star, I was sure the error was in my addition of the ```wsus``` functionality to the ```win_pkg.py``` file. 

BUT thanks to you, jsut realised that the ```osversion``` is actually a range that applies it is NOT equalto ```6.3``` or ```6.2``` or ```6.1```

it needs to be something like 
for ```6.3``` ```osversion >= 6.3 AND < 6.4```
for ```6.2``` ```osversion >= 6.2 AND < 6.3```
for ```6.1``` ```osversion >= 6.1 AND < 6.2``` 

I got it! 

```
{% if salt['pkg.compare_versions'](grains['osversion'], '==', '6.2') and salt['pkg.compare_versions'](grains['osversion'], '<', '6.3') %}
    full_name: 'Windows Management Framework 5.0 Production Preview (KB3066438)'
    {% elif salt['pkg.compare_versions'](grains['osversion'], '==', '6.1') and salt['pkg.compare_versions'](grains['osversion'], '<', '6.2') %}
    full_name: 'Windows Management Framework 5.0 Production Preview (KB3066439)'
    {% elif salt['pkg.compare_versions'](grains['osversion'], '>=', '6.3') and salt['pkg.compare_versions'](grains['osversion'], '<', '6.4') %}
    full_name: 'Windows Management Framework 5.0 Production Preview (KB3066437)'
```

Thanks Alexander! Learned a couple new tricks from you.

**Comment by TheBigBear** [comment]
@twangboy this https://github.com/saltstack/salt/pull/27224 is now tested and works. any chance this can even be merged back into 2015.8 branch maybe for 2015.8.1 (or before?)
